In [70]:
import pandas as pd

In [71]:
import os

FLOODING_311_CACHE = 'flooding_311_2025.csv'
FORCE_REFRESH = False  # set to True to force a fresh pull from NYC Open Data

if os.path.exists(FLOODING_311_CACHE) and not FORCE_REFRESH:
    flooding_311 = pd.read_csv(FLOODING_311_CACHE)
    print(f"Loaded cached {FLOODING_311_CACHE} ({len(flooding_311)} rows)")
else:
    flooding_311 = pd.read_json('https://data.cityofnewyork.us/resource/erm2-nwe9.json?$query=SELECT%0A%20%20%60unique_key%60%2C%0A%20%20%60created_date%60%2C%0A%20%20%60closed_date%60%2C%0A%20%20%60agency%60%2C%0A%20%20%60agency_name%60%2C%0A%20%20%60complaint_type%60%2C%0A%20%20%60descriptor%60%2C%0A%20%20%60descriptor_2%60%2C%0A%20%20%60location_type%60%2C%0A%20%20%60incident_zip%60%2C%0A%20%20%60incident_address%60%2C%0A%20%20%60street_name%60%2C%0A%20%20%60cross_street_1%60%2C%0A%20%20%60cross_street_2%60%2C%0A%20%20%60intersection_street_1%60%2C%0A%20%20%60intersection_street_2%60%2C%0A%20%20%60address_type%60%2C%0A%20%20%60city%60%2C%0A%20%20%60landmark%60%2C%0A%20%20%60facility_type%60%2C%0A%20%20%60status%60%2C%0A%20%20%60due_date%60%2C%0A%20%20%60resolution_description%60%2C%0A%20%20%60resolution_action_updated_date%60%2C%0A%20%20%60community_board%60%2C%0A%20%20%60council_district%60%2C%0A%20%20%60police_precinct%60%2C%0A%20%20%60bbl%60%2C%0A%20%20%60borough%60%2C%0A%20%20%60x_coordinate_state_plane%60%2C%0A%20%20%60y_coordinate_state_plane%60%2C%0A%20%20%60open_data_channel_type%60%2C%0A%20%20%60park_facility_name%60%2C%0A%20%20%60park_borough%60%2C%0A%20%20%60vehicle_type%60%2C%0A%20%20%60taxi_company_borough%60%2C%0A%20%20%60taxi_pick_up_location%60%2C%0A%20%20%60bridge_highway_name%60%2C%0A%20%20%60bridge_highway_direction%60%2C%0A%20%20%60road_ramp%60%2C%0A%20%20%60bridge_highway_segment%60%2C%0A%20%20%60latitude%60%2C%0A%20%20%60longitude%60%2C%0A%20%20%60location%60%0AWHERE%0A%20%20%60created_date%60%0A%20%20%20%20BETWEEN%20%222025-01-01T00%3A00%3A00%22%20%3A%3A%20floating_timestamp%0A%20%20%20%20AND%20%222025-12-31T23%3A45%3A00%22%20%3A%3A%20floating_timestamp%0A%20%20AND%20caseless_one_of(%0A%20%20%20%20%60descriptor%60%2C%0A%20%20%20%20%22Flooding%20on%20Highway%22%2C%0A%20%20%20%20%22Flooded%22%2C%0A%20%20%20%20%22Flooding%20on%20Street%22%2C%0A%20%20%20%20%22Street%20Flooding%20(SJ)%22%2C%0A%20%20%20%20%22Highway%20Flooding%20(SH)%22%2C%0A%20%20%20%20%22Ponding%22%2C%0A%20%20%20%20%22Puddle%20in%20Ground%22%2C%0A%20%20%20%20%22Puddle%20on%20Driveway%22%2C%0A%20%20%20%20%22Puddle%20on%20Sidewalk%22%2C%0A%20%20%20%20%22Catch%20Basin%20Clogged%22%2C%0A%20%20%20%20%22Catch%20Basin%20Clogged%2FFlooding%20(Use%20Comments)%20(SC)%22%2C%0A%20%20%20%20%22Manhole%20Overflow%20(Use%20Comments)%20(SA1)%22%2C%0A%20%20%20%20%22Backup%22%2C%0A%20%20%20%20%22Sewer%20Backup%20(Use%20Comments)%20(SA)%22%2C%0A%20%20%20%20%22Drain%2FPipe%20Clogged%22%2C%0A%20%20%20%20%22Sewer%20or%20Drain%22%2C%0A%20%20%20%20%22Water%20or%20Sewer%20Runoff%22%2C%0A%20%20%20%20%22Water%20in%20Basement%22%0A%20%20)%0AORDER%20BY%20%60created_date%60%20DESC%20NULL%20FIRST%0ALIMIT%2050000')
    flooding_311.to_csv(FLOODING_311_CACHE, index=False)
    print(f"Fetched from NYC Open Data and cached to {FLOODING_311_CACHE} ({len(flooding_311)} rows)")


Loaded cached flooding_311_2025.csv (19617 rows)


C:\Users\sriva\AppData\Local\Temp\ipykernel_6444\4151326925.py:7: DtypeWarning: Columns (0: location_type, 1: landmark) have mixed types. Specify dtype option on import or set low_memory=False.
  flooding_311 = pd.read_csv(FLOODING_311_CACHE)


In [72]:
display(flooding_311.head())
flooding_311.shape

,unique_key,created_date,closed_date,agency,agency_name,complaint_type,descriptor,descriptor_2,incident_zip,incident_address,...,park_facility_name,park_borough,latitude,longitude,location,intersection_street_1,intersection_street_2,location_type,landmark,facility_type
0,67349010,2025-12-31T22:29:00.000,2026-01-01T03:00:00.000,DEP,Department of Environmental Protection,Sewer,Sewer Backup (Use Comments) (SA),SA,10306.0,18 CELTIC PLACE,...,Unspecified,STATEN ISLAND,40.565637,-74.105791,"{'type': 'Point', 'coordinates': [-74.10579125...",NaN,NaN,NaN,NaN,NaN
1,67351391,2025-12-31T20:26:00.000,2026-01-01T05:45:00.000,DEP,Department of Environmental Protection,Sewer,Sewer Backup (Use Comments) (SA),SA,11412.0,119-48 194 STREET,...,Unspecified,QUEENS,40.689335,-73.755965,"{'type': 'Point', 'coordinates': [-73.75596487...",NaN,NaN,NaN,NaN,NaN
2,67344182,2025-12-31T19:35:00.000,2025-12-31T21:50:00.000,DEP,Department of Environmental Protection,Sewer,Sewer Backup (Use Comments) (SA),SA,11432.0,169-02 HILLSIDE AVENUE,...,Unspecified,QUEENS,40.710762,-73.792869,"{'type': 'Point', 'coordinates': [-73.79286947...",NaN,NaN,NaN,NaN,NaN
3,67347791,2025-12-31T18:53:00.000,2025-12-31T19:55:00.000,DEP,Department of Environmental Protection,Sewer,Sewer Backup (Use Comments) (SA),SA,11234.0,1145 EAST 52 STREET,...,Unspecified,BROOKLYN,40.632012,-73.926312,"{'type': 'Point', 'coordinates': [-73.92631237...",NaN,NaN,NaN,NaN,NaN
4,67349011,2025-12-31T17:41:00.000,2026-01-01T01:20:00.000,DEP,Department of Environmental Protection,Sewer,Sewer Backup (Use Comments) (SA),SA,11412.0,119-48 191 STREET,...,Unspecified,QUEENS,40.688429,-73.758514,"{'type': 'Point', 'coordinates': [-73.75851390...",NaN,NaN,NaN,NaN,NaN


(19617, 36)

In [73]:
flooding_311['created_date'] = pd.to_datetime(flooding_311['created_date'])
flooding_311['date'] = flooding_311['created_date'].dt.date

In [ ]:
sensors = pd.read_csv('../../final_deployed_sensors.csv')[['sensor_name', 'latitude', 'longitude']]

qc = pd.read_csv('../FloodNet_Data_Manual_QC.csv')
qc['date'] = pd.to_datetime(qc['Date']).dt.date

qc_long = qc.assign(Sensors=qc['Sensors'].str.split(',\s*')).explode('Sensors')
qc_long['sensor_name'] = qc_long['Sensors'].str.strip()
qc_long = qc_long.merge(sensors, on='sensor_name', how='left')

In [75]:
reports = flooding_311[['date', 'latitude', 'longitude', 'incident_address', 'descriptor']].dropna(subset=['latitude', 'longitude']).copy()
reports['latitude'] = reports['latitude'].astype(float)
reports['longitude'] = reports['longitude'].astype(float)

In [76]:
import numpy as np

def haversine_miles(lat1, lon1, lat2, lon2):
    lat1, lon1, lat2, lon2 = map(np.radians, [lat1, lon1, lat2, lon2])
    dlat, dlon = lat2 - lat1, lon2 - lon1
    a = np.sin(dlat / 2)**2 + np.cos(lat1) * np.cos(lat2) * np.sin(dlon / 2)**2
    return 2 * 3958.8 * np.arcsin(np.sqrt(a))

In [77]:
RADIUS_MILES = 0.5

def nearest_report(row):
    if pd.isna(row['latitude']):
        return pd.Series({'nearest_311_miles': np.nan, 'nearest_311_address': None, 'nearest_311_descriptor': None})
    same_day = reports[reports['date'] == row['date']]
    if same_day.empty:
        return pd.Series({'nearest_311_miles': np.nan, 'nearest_311_address': None, 'nearest_311_descriptor': None})
    dists = haversine_miles(row['latitude'], row['longitude'], same_day['latitude'], same_day['longitude'])
    idx = dists.idxmin()
    return pd.Series({
        'nearest_311_miles': dists.loc[idx],
        'nearest_311_address': same_day.loc[idx, 'incident_address'],
        'nearest_311_descriptor': same_day.loc[idx, 'descriptor'],
    })

qc_long[['nearest_311_miles', 'nearest_311_address', 'nearest_311_descriptor']] = qc_long.apply(nearest_report, axis=1)
qc_long['corroborated'] = qc_long['nearest_311_miles'] <= RADIUS_MILES

def nearby_reports(row):
    if pd.isna(row['latitude']):
        return []
    same_day = reports[reports['date'] == row['date']]
    if same_day.empty:
        return []
    dists = haversine_miles(row['latitude'], row['longitude'], same_day['latitude'], same_day['longitude'])
    return same_day.loc[dists <= RADIUS_MILES, 'incident_address'].tolist()

qc_long['nearby_311_addresses'] = qc_long.apply(nearby_reports, axis=1)
qc_long['num_nearby_311'] = qc_long['nearby_311_addresses'].str.len()

display(qc_long.head())
print(qc_long['corroborated'].value_counts())

,Date,Count,Weather,Sensors,date,sensor_name,latitude,longitude,nearest_311_miles,nearest_311_address,nearest_311_descriptor,corroborated,nearby_311_addresses,num_nearby_311
0,01/12/2025,1,NaN,SI - McLaughlin St/Agnes Pl,2025-01-12,SI - McLaughlin St/Agnes Pl,40.589132,-74.072692,3.526754,10 DEWHURST STREET,Sewer Backup (Use Comments) (SA),False,[],0
1,01/14/2025,1,NaN,SI - Grimsby St/ Mapleton Ave,2025-01-14,SI - Grimsby St/ Mapleton Ave,40.574836,-74.093448,0.722593,211 FREMONT AVENUE,Sewer Backup (Use Comments) (SA),False,[],0
2,01/15/2025,1,NaN,SI - Grimsby St/ Mapleton Ave,2025-01-15,SI - Grimsby St/ Mapleton Ave,40.574836,-74.093448,2.131742,NaN,Manhole Overflow (Use Comments) (SA1),False,[],0
3,01/16/2025,1,NaN,Q - Beach Channel Dr/Beach 48th St,2025-01-16,Q - Beach Channel Dr/Beach 48th St,40.595001,-73.779035,1.636680,16-16 REDFERN AVENUE,Sewer Backup (Use Comments) (SA),False,[],0
4,01/18/2025,1,NaN,SI - Grimsby St/ Mapleton Ave,2025-01-18,SI - Grimsby St/ Mapleton Ave,40.574836,-74.093448,0.048685,771 NUGENT AVENUE,Street Flooding (SJ),True,[771 NUGENT AVENUE],1


corroborated
False    213
True      14
Name: count, dtype: int64


## Bringing in rain and tide/surge corroboration

311 alone is a weak signal (biased toward denser, higher-reporting areas — see above).
`Tidal_Corroboration.ipynb` already built a `rain_flag` (rain at the sensor's nearest
weather station, day-of or day-before) and `tide_class` (whether that day's high tide
was unusually high astronomically, surged, both, or neither) for every QC event. Merge
that in here instead of recomputing it, so each bump can be judged on rain, tide/surge,
and 311 together.

One row — `SI - Grimsby St/Mapleton Ave` on 2025-12-19 — is missing from that file: it
was already flagged there as a QC slip (actually in the public feed on time, not a real
gap), so its tidal columns will come back NaN. That's expected, not a merge bug.

In [ ]:
tidal_events = pd.read_csv('../3_tidal/tidal_corroboration_events.csv', parse_dates=['date'])
tidal_events['date'] = tidal_events['date'].dt.date

qc_long = qc_long.merge(
    tidal_events[['date', 'sensor_name', 'rain_flag', 'tide_class', 'astro_pctile', 'surge_ft_h']],
    on=['date', 'sensor_name'], how='left', indicator='_tidal_merge'
)

qc_long['known_qc_slip'] = (
    (qc_long['sensor_name'] == 'SI - Grimsby St/ Mapleton Ave')
    & (qc_long['date'] == pd.Timestamp('2025-12-19').date())
)

# anything unmatched that ISN'T the known slip would be a real problem worth investigating
unexpected_unmatched = qc_long[(qc_long['_tidal_merge'] == 'left_only') & ~qc_long['known_qc_slip']]
print(f"unexpected unmatched rows: {len(unexpected_unmatched)}")
qc_long = qc_long.drop(columns='_tidal_merge')

## Three signals, side by side

Rain and tide/surge don't apply to the same sensors — a pluvial (street/sewer) sensor
isn't expected to show a tide signal, and a coastal sensor isn't expected to show rain.
So instead of averaging the three checks into one score, keep them as separate columns
and surface only the bumps where **none** of the three signals fired — those are the
ones with no independent corroboration at all, and worth a second look (or acceptance
that FloodNet caught something none of the other signals were positioned to see).

In [79]:
qc_long['tide_signal'] = qc_long['tide_class'].notna() & (qc_long['tide_class'] != 'neither')
qc_long['num_signals'] = (
    qc_long['rain_flag'].fillna(False).astype(bool).astype(int)
    + qc_long['tide_signal'].astype(int)
    + qc_long['corroborated'].fillna(False).astype(bool).astype(int)
)

summary_cols = ['Date', 'sensor_name', 'rain_flag', 'tide_class', 'num_nearby_311', 'corroborated',
                 'nearest_311_miles', 'nearest_311_descriptor',
                 'num_signals', 'known_qc_slip']
display(qc_long[summary_cols].sort_values('num_signals'))

print(qc_long['num_signals'].value_counts().sort_index())
print()
no_signal = qc_long[(qc_long['num_signals'] == 0) & ~qc_long['known_qc_slip']]
print(f"{len(no_signal)} bumps have zero corroboration across rain, tide/surge, and nearby 311:")
display(no_signal[summary_cols])

,Date,sensor_name,rain_flag,tide_class,num_nearby_311,corroborated,nearest_311_miles,nearest_311_descriptor,num_signals,known_qc_slip
0,01/12/2025,SI - McLaughlin St/Agnes Pl,False,neither,0,False,3.526754,Sewer Backup (Use Comments) (SA),0,False
1,01/14/2025,SI - Grimsby St/ Mapleton Ave,False,neither,0,False,0.722593,Sewer Backup (Use Comments) (SA),0,False
2,01/15/2025,SI - Grimsby St/ Mapleton Ave,False,neither,0,False,2.131742,Manhole Overflow (Use Comments) (SA1),0,False
3,01/16/2025,Q - Beach Channel Dr/Beach 48th St,False,neither,0,False,1.636680,Sewer Backup (Use Comments) (SA),0,False
6,01/19/2025,SI - McLaughlin St/Agnes Pl,False,neither,0,False,5.435552,Sewer Backup (Use Comments) (SA),0,False
...,...,...,...,...,...,...,...,...,...,...
157,07/16/2025,SI - Bedford Ave/Kiswick St,True,neither,1,True,0.298858,Street Flooding (SJ),2,False
158,07/16/2025,Q - Beach Channel Dr/Beach 48th St,True,neither,1,True,0.446304,Catch Basin Clogged/Flooding (Use Comments) (SC),2,False
184,11/08/2025,Q - Beach Channel Dr/Beach 48th St,True,tidal (high astronomical),1,True,0.186640,Sewer Backup (Use Comments) (SA),3,False
180,10/30/2025,SI - Minthorne St/ Victory Blvd,True,surge,2,True,0.316890,Street Flooding (SJ),3,False


num_signals
0    127
1     69
2     28
3      3
Name: count, dtype: int64

126 bumps have zero corroboration across rain, tide/surge, and nearby 311:


,Date,sensor_name,rain_flag,tide_class,num_nearby_311,corroborated,nearest_311_miles,nearest_311_descriptor,num_signals,known_qc_slip
0,01/12/2025,SI - McLaughlin St/Agnes Pl,False,neither,0,False,3.526754,Sewer Backup (Use Comments) (SA),0,False
1,01/14/2025,SI - Grimsby St/ Mapleton Ave,False,neither,0,False,0.722593,Sewer Backup (Use Comments) (SA),0,False
2,01/15/2025,SI - Grimsby St/ Mapleton Ave,False,neither,0,False,2.131742,Manhole Overflow (Use Comments) (SA1),0,False
3,01/16/2025,Q - Beach Channel Dr/Beach 48th St,False,neither,0,False,1.636680,Sewer Backup (Use Comments) (SA),0,False
6,01/19/2025,SI - McLaughlin St/Agnes Pl,False,neither,0,False,5.435552,Sewer Backup (Use Comments) (SA),0,False
...,...,...,...,...,...,...,...,...,...,...
215,12/26/2025,BX - Watson Ave/Close Ave,False,neither,0,False,0.846322,Catch Basin Clogged/Flooding (Use Comments) (SC),0,False
223,12/29/2025,SI - Grimsby St/ Mapleton Ave,False,neither,0,False,2.392247,Sewer Backup (Use Comments) (SA),0,False
224,12/30/2025,SI - Grimsby St/ Mapleton Ave,False,neither,0,False,0.854847,Sewer Backup (Use Comments) (SA),0,False
225,12/31/2025,SI - Bedford Ave/Kiswick St,False,neither,0,False,0.760786,Sewer Backup (Use Comments) (SA),0,False


In [80]:
two_of_three = qc_long[qc_long['num_signals'] == 2]
print(f"{len(two_of_three)} bumps have 2 of 3 signals corroborating:")
display(two_of_three[summary_cols])

28 bumps have 2 of 3 signals corroborating:


,Date,sensor_name,rain_flag,tide_class,num_nearby_311,corroborated,nearest_311_miles,nearest_311_descriptor,num_signals,known_qc_slip
33,02/13/2025,BX - Watson Ave/Close Ave,True,surge,0,False,0.766725,Sewer Backup (Use Comments) (SA),2,False
44,03/06/2025,SI - Baden Pl/ Mapleton Ave,True,surge,0,False,0.607436,Street Flooding (SJ),2,False
45,03/06/2025,BX - Watson Ave/Close Ave,True,surge,0,False,0.913491,Catch Basin Clogged/Flooding (Use Comments) (SC),2,False
61,03/19/2025,SI - Grimsby St/ Mapleton Ave,True,neither,1,True,0.061465,Sewer Backup (Use Comments) (SA),2,False
62,03/21/2025,SI - Hylan Blvd/ Jefferson Blvd,True,neither,1,True,0.138182,Manhole Overflow (Use Comments) (SA1),2,False
63,03/21/2025,BX - Tier St/William Ave,True,surge,0,False,2.072608,Catch Basin Clogged/Flooding (Use Comments) (SC),2,False
71,03/29/2025,BX - Watson Ave/Close Ave,True,tidal (high astronomical),0,False,3.734945,Catch Basin Clogged/Flooding (Use Comments) (SC),2,False
72,03/30/2025,BX - Watson Ave/Close Ave,True,tidal (high astronomical),0,False,2.309497,Sewer Backup (Use Comments) (SA),2,False
76,03/31/2025,SI - Baden Pl/ Mapleton Ave,False,tidal (high astronomical),1,True,0.268673,Catch Basin Clogged/Flooding (Use Comments) (SC),2,False
77,03/31/2025,SI - Grimsby St/ Mapleton Ave,False,tidal (high astronomical),1,True,0.358610,Catch Basin Clogged/Flooding (Use Comments) (SC),2,False


## Mapping corroboration: where and when

`num_signals` (0-3) is a magnitude, not a category, so both views below use one
sequential blue ramp (light = fewer signals, dark = more) instead of arbitrary colors.

- **Static map** — one marker per sensor (11 total), colored by that sensor's *mean*
  `num_signals` across all its bumps, sized by how many bumps it had. Shows which
  sensors tend to be well- vs. poorly-corroborated overall.
- **Slider map** — every individual bump plotted at its sensor's location, with a
  time slider (via `folium.plugins.TimestampedGeoJson`) stepping through 2025 day by
  day. Points accumulate as the slider advances, so dragging it shows *when* each
  sensor's bumps happened and whether corroboration clusters in time (e.g. a wet
  season) or is scattered — the same "where vs. when" split as a static map + chart,
  but keeping both dimensions inside folium since that's the renderer that's
  actually working in this notebook.

In [81]:
sensor_summary = qc_long.groupby('sensor_name').agg(
    n_bumps=('num_signals', 'size'),
    mean_signals=('num_signals', 'mean'),
    latitude=('latitude', 'first'),
    longitude=('longitude', 'first'),
).reset_index()

display(sensor_summary.sort_values('mean_signals'))

,sensor_name,n_bumps,mean_signals,latitude,longitude
6,SI - Grimsby St/ Mapleton Ave,40,0.375000,40.574836,-74.093448
10,SI - Snug Harbor Rd / Kissel Ave,13,0.384615,40.644221,-74.106407
8,SI - McLaughlin St/Agnes Pl,34,0.500000,40.589132,-74.072692
7,SI - Hylan Blvd/ Jefferson Blvd,9,0.555556,40.581063,-74.098494
9,SI - Minthorne St/ Victory Blvd,36,0.583333,40.637624,-74.075250
3,Q - Beach Channel Dr/Beach 48th St,26,0.615385,40.595001,-73.779035
4,SI - Baden Pl/ Mapleton Ave,18,0.666667,40.573189,-74.090386
1,BX - Watson Ave/Close Ave,18,0.777778,40.825626,-73.882366
2,Q - Beach 49th St/Rockaway Beach Blvd,11,0.818182,40.593339,-73.779743
0,BX - Tier St/William Ave,14,0.857143,40.848094,-73.789491


In [82]:
import folium
import branca.colormap as cm

colormap = cm.linear.Blues_09.scale(0, 3)
colormap.caption = 'Mean # corroborating signals (0-3)'

m = folium.Map(
    location=[sensor_summary['latitude'].mean(), sensor_summary['longitude'].mean()],
    zoom_start=10, tiles='cartodbpositron'
)

for _, row in sensor_summary.iterrows():
    folium.CircleMarker(
        location=[row['latitude'], row['longitude']],
        radius=6 + row['n_bumps'] / 4,
        color=colormap(row['mean_signals']),
        fill=True,
        fill_color=colormap(row['mean_signals']),
        fill_opacity=0.85,
        weight=1,
        tooltip=f"{row['sensor_name']} | {row['n_bumps']} bumps | mean signals: {row['mean_signals']:.2f}"
    ).add_to(m)

colormap.add_to(m)
m

In [83]:
from folium.plugins import TimestampedGeoJson

def signals_fired(row):
    fired = []
    if bool(row['rain_flag']) if pd.notna(row['rain_flag']) else False:
        fired.append('rain')
    if row['tide_signal']:
        fired.append(f"tide/surge ({row['tide_class']})")
    if row['corroborated']:
        fired.append(f"311 nearby ({row['num_nearby_311']})")
    return ', '.join(fired) if fired else 'none'

def to_feature(row):
    color = colormap(row['num_signals'])
    return {
        'type': 'Feature',
        'geometry': {'type': 'Point', 'coordinates': [row['longitude'], row['latitude']]},
        'properties': {
            'time': row['date'].isoformat(),
            'popup': (
                f"{row['sensor_name']}<br>{row['Date']}<br>"
                f"signals fired ({row['num_signals']}/3): {signals_fired(row)}"
            ),
            'icon': 'circle',
            'iconstyle': {'fillColor': color, 'color': color, 'fillOpacity': 0.8, 'radius': 6},
        },
    }

bump_points = qc_long.dropna(subset=['latitude', 'longitude'])
features = bump_points.apply(to_feature, axis=1).tolist()

m2 = folium.Map(
    location=[bump_points['latitude'].mean(), bump_points['longitude'].mean()],
    zoom_start=10, tiles='cartodbpositron'
)

TimestampedGeoJson(
    {'type': 'FeatureCollection', 'features': features},
    period='P1D',
    add_last_point=True,
    auto_play=False,
).add_to(m2)

colormap.add_to(m2)
m2

## Conclusion

100/227 (44%) of the manually-flagged bumps have at least one independent corroborating
signal — rain, an unusually high/surging tide, or a nearby same-day 311 call — and
31/227 have two or three. That's evidence these bumps aren't noise: they're real
weather/tide events lining up with what FloodNet logged.

The 127 with zero signals (126 once the one known QC slip is set aside) aren't
evidence of the opposite. Rain doesn't apply to a coastal sensor and tide/surge
doesn't apply to a pluvial one, so a sensor only ever has access to one real physical
signal plus a weak, reporting-biased 311 check — three checks were never going to
cover every sensor. Zero signals most often means "these particular checks weren't
positioned to see it," not "this didn't happen."

Combined with the independent feed-verification pass in `Weather_Data_Scrape.ipynb`
(62 of 63 rain-driven QC events confirmed genuinely absent from the public feed, only
1 flagged as a QC slip and not a real gap), I'm treating the 2025 data gap as real:
these are actual floods missing from FloodNet's public dataset, not manual-QC
false positives.